Incorrect behavior of "cursor: auto" over links

vitorroriz opened the pull request:
#### b5d767e418f3d45709b214a545d9a04076954823
<pre>
Incorrect behavior of &quot;cursor: auto&quot; over links
<a href="https://bugs.webkit.org/show_bug.cgi?id=173909">https://bugs.webkit.org/show_bug.cgi?id=173909</a>
rdar://99495893

Reviewed by NOBODY (OOPS!).

The initial attempt to fix this bug and comply to spec was made
at <a href="https://commits.webkit.org/253685@main">https://commits.webkit.org/253685@main</a>

After that, a:any-link no longer depends on &apos;cursor: auto&apos;
for achieving &quot;Hand&quot; (css: pointer) cursor. This is because, according to spec,
&apos;cursor: auto&apos; can only result in default or text cursor. Therefore, 253685@main
sets &apos;cursor: pointer&apos; in the UA stylesheet (html.css).

Because of that, a test case for cursor types for recognized text (live-text)
was broken. There, we were testing a &lt;img&gt; nested in a &lt;a href=&quot;&quot;&gt;. Since,
this &lt;a&gt; is now selected to have &apos;cursor: pointer&apos;, its nested &lt;img&gt;
has no longer &apos;cursor: auto&apos;, which is required for resulting in IBeam (text)
cursor for the overlay installed in the img (see EventHandler::selectCursor).

Therefore, we are updating the test case here. A &lt;img&gt; inside a &lt;a href=&quot;...&quot;&gt;
is by default styled with cursor: pointer, and therefore hovering over
its overlay doesn&apos;t show IBeam (text) cursor, unless user styles its cursor to auto.

* LayoutTests/fast/images/text-recognition/mac/cursor-types-for-recognized-text-expected.txt:
* LayoutTests/fast/images/text-recognition/mac/cursor-types-for-recognized-text.html:
Updating tests. Note that one of the test cases is commented out. As stated in the comment,
this would fail due to <a href="https://bugs.webkit.org/show_bug.cgi?id=244944">https://bugs.webkit.org/show_bug.cgi?id=244944</a> (radar: 99495849).
In short, EventHandler::selectCursor will select an IBeam cursor although it shouldn&apos;t.
This is because Node::canStartSelection() is not false for user-select: none.
* LayoutTests/platform/mac-wk2/TestExpectations:
Reseting TextExpectations for this test, before set as failure.
* Source/WebCore/html/shadow/imageOverlay.css:
(div#image-overlay):
We should not style cursor for the live-text overlay that is installed
on top of images with text. Doing that would disable user to control which
cursor is displayed for the text portion of the image, since the overlay
is in the shadow tree, hosted by the image element. I.e.: user&apos;s cursor style
for the image would not be propagated to the overlay.
</pre>

nt1m commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -1459,7 +1459,7 @@ std::optional<Cursor> EventHandler::selectCursor(const HitTestResult& result, bo
 
     auto renderer = node->renderer();
-    auto* style = renderer ? &renderer->style() : nullptr;
+    auto* style = renderer ? &renderer->style() : node->computedStyle();
```
auto* style = node->computedStyle();

Also worth a comment describing why the computed style is used and not the renderStyle like it usually is the case.

vitorroriz replied:
Can we always use the computed style here, not only when the node has no renderer?

nt1m replied:
I think so, @anttijk wdyt?

vitorroriz replied:
He DM me that it "should be ok"

nt1m commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -1459,7 +1459,7 @@ std::optional<Cursor> EventHandler::selectCursor(const HitTestResult& result, bo
         return std::nullopt;
 
     auto renderer = node->renderer();
```
Can you move this variable further down closer to where it's used? (also it should be `auto*`)

nt1m commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -1459,7 +1459,7 @@ std::optional<Cursor> EventHandler::selectCursor(const HitTestResult& result, bo
 
     auto renderer = node->renderer();
-    auto* style = renderer ? &renderer->style() : nullptr;
+    auto* style = node->computedStyle();
```
Can you add a comment explaining why `computedStyle()` is appropriate here?

vitorroriz committed:
Incorrect behavior of "cursor: auto" over links
https://bugs.webkit.org/show_bug.cgi?id=173909
rdar://33054544

Reviewed by Simon Fraser.

* LayoutTests/imported/w3c/web-platform-tests/html/rendering/the-css-user-agent-style-sheet-and-presentational-hints/mouse-cursor-imagemap-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/the-css-user-agent-style-sheet-and-presentational-hints/no-help-cursor-on-links.historical-expected.txt:
* Source/WebCore/css/html.css:
hand cursor (pointer) should not be rendered through 'auto' behavior of agent.
(a:any-link):
(a:any-link:read-write):
(area:any-link):
* Source/WebCore/page/EventHandler.cpp:
Removing hand cursor calculation from 'auto' case. Auto should just result in default or text cursors.
We are now using computedStyle instead of renderer->style() because we still need to access style
for elements without a renderer, like <area>.
(WebCore::EventHandler::updateCursor):
(WebCore::EventHandler::selectCursor):
(WebCore::EventHandler::handleMouseMoveEvent):
* Source/WebCore/page/EventHandler.h:
Since we don't handle hand cursor anymore on selectCursor, selectCursor and updateCursor
no longer need to receive shiftKey.

Canonical link: https://commits.webkit.org/253685@main
